src/cpu/testers/traffic_gen/gups_gen.cc

```diff
@@ -0,0 +1,321 @@
+#include "cpu/testers/traffic_gen/gups_gen.hh"
+
+#include <cstring>
+#include <string>
+
+#include "base/random.hh"
+#include "debug/GUPSGen.hh"
+#include "sim/sim_exit.hh"
+
+#define POLY 0x0000000000000007UL
+#define PERIOD 1317624576693539401L
+
+
+GUPSGen::GUPSGen(const GUPSGenParams &params):
+    ClockedObject(params),
+    system(params.system),
+    requestorId(system->getRequestorId(this)),
+    nextGenEvent([this]{ generateNextReq(); }, name()),
+    nextSendEvent([this]{ sendNextBatch(); }, name()),
+    port(name() + ".port", this),
+    startAddr(params.start_addr),
+    memSize(params.mem_size),
+    blockSize(params.block_size),
+    doneReading(false),
+    onTheFlyRequests(0)
+{}
+
+Port&
+GUPSGen::getPort(const std::string &if_name, PortID idx)
+{
+    if (if_name != "port"){
+        return ClockedObject::getPort(if_name, idx);
+    }
+    else{
+        return port;
+    }
+}
+
+void
+GUPSGen::startup()
+{
+
+    uint64_t stride_size = blockSize / sizeof(uint64_t);
+
+    for (uint64_t start_index = 0; start_index < tableSize;
+                                start_index += stride_size)
+    {
+        uint8_t write_data[blockSize];
+        for (uint64_t offset = 0; offset < stride_size; offset++)
+        {
+            uint64_t value = start_index + offset;
+            std::memcpy(write_data + offset * sizeof(uint64_t),
+                        &value, sizeof(uint64_t));
+        }
+        Addr addr = indexToAddr(start_index);
+        PacketPtr pkt = getWritePacket(addr, blockSize, write_data);
+        port.sendFunctionalPacket(pkt);
+    }
+
+    /*
+     * initialize memory by functional access
+     * call generateNextBatch here.
+    */
+   schedule(nextGenEvent, curTick());
+}
+
+void
+GUPSGen::init()
+{
+
+    tableSize = memSize / sizeof(uint64_t);
+    numUpdates = 4 * tableSize;
+
+    for (int i = 0; i < 128; i++)
+    {
+        randomSeeds[i] = randomNumberGenerator((numUpdates / 128) * i);
+    }
+
+}
+
+void
+GUPSGen::generateNextReq()
+{
+    static uint64_t iterations = 0;
+    static int rIndex = 0;
+
+    assert (rIndex < 128);
+
+    if (iterations >= numUpdates / 128){
```

Comment:
what is `128`?

Reply:
The original code does updates in batches of 128 elements in the Table.

Reply:
Where is the "original code"? What is the "original code"? What is the "table"?

Make sure that when you write code there is enough context for someone else to read it and understand it.